fault_proof/bin/proposer.rs

```diff
@@ -0,0 +1,327 @@
+                self.try_resolve_unchallenged_game(index).await?;
+            }
+        } else {
+            tracing::info!(
+                "Oldest game {:?} at index {:?} is not resolved, not attempting resolution",
+                game_address,
+                oldest_game_index
+            );
+        }
+
+        Ok(())
+    }
+
+    /// Runs the proposer indefinitely.
+    async fn run(&self) -> Result<()> {
+        tracing::info!("OP Succinct Proposer running...");
+        let mut interval = time::interval(Duration::from_secs(self.config.fetch_interval));
+
+        loop {
+            interval.tick().await;
+
+            let _span = tracing::info_span!("[[Proposing]]").entered();
+
+            let safe_l2_head_block_number = self
+                .l2_provider
+                .get_l2_block_by_number(BlockNumberOrTag::Safe)
+                .await?
+                .header
+                .number;
+            tracing::debug!("Safe L2 head block number: {:?}", safe_l2_head_block_number);
+
+            let latest_valid_proposal = self
+                .factory
+                .get_latest_valid_proposal(self.l1_provider.clone(), self.l2_provider.clone())
+                .await?;
+
+            let (next_l2_block_number_for_proposal, parent_game_index) = match latest_valid_proposal
+            {
+                Some((latest_block, latest_game_idx)) => (
+                    latest_block + U256::from(self.config.proposal_interval_in_blocks),
+                    latest_game_idx.to::<u32>(),
+                ),
+                None => {
+                    // For first game, start from genesis L2 block number + proposal interval
+                    let genesis_l2_block_number = self
+                        .factory
+                        .get_genesis_l2_block_number(
+                            self.config.game_type,
+                            self.l1_provider.clone(),
+                        )
+                        .await?;
+
+                    (
+                        genesis_l2_block_number
+                            .checked_add(U256::from(self.config.proposal_interval_in_blocks))
+                            .unwrap(),
+                        u32::MAX, // Use max value for first game's parent index
+                    )
+                }
+            };
+
+            if U256::from(safe_l2_head_block_number) > next_l2_block_number_for_proposal {
+                self.create_game(next_l2_block_number_for_proposal, parent_game_index)
+                    .await?;
+            }
+
+            drop(_span);
+
+            // Only attempt game resolution if enabled
+            if self.config.enable_game_resolution {
+                let _span = tracing::info_span!("[[Resolving]]").entered();
+
+                if let Err(e) = self.resolve_unchallenged_games().await {
+                    tracing::warn!("Failed to resolve unchallenged games: {:?}", e);
+                }
+
+                drop(_span);
```

Comment:
This feels like an anti-pattern. Is there a cleaner way to separate the logging for the proposing/resolving sections?